iceoryx_hoofs/include/iceoryx_hoofs/cxx/span.hpp

```diff
@@ -0,0 +1,414 @@
+// Copyright (c) 2021 - 2022 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+#ifndef IOX_HOOFS_CXX_SPAN_HPP
+#define IOX_HOOFS_CXX_SPAN_HPP
+
+#include "iceoryx_hoofs/cxx/requires.hpp"
+#include "iceoryx_hoofs/cxx/span_iterator.hpp"
+#include "iceoryx_hoofs/cxx/type_traits.hpp"
+#include "iox/uninitialized_array.hpp"
+
+#include <algorithm>
+#include <cstdint>
+#include <cstdio>
+#include <limits>
+#include <utility>
+
+using namespace std;
+
+namespace iox
+{
+namespace cxx
+{
+// constants
+constexpr uint64_t dynamic_extent = std::numeric_limits<uint64_t>::max();
+
+template <class From, class To>
+constexpr bool is_convertible_v = std::is_convertible<From, To>::value;
+
+// Implementation of C++17 std::size() and std::data()
+template <class C>
+constexpr auto size(const C& c) -> decltype(c.size())
+{
+    return c.size();
+}
+
+template <class T, std::uint64_t N>
+constexpr std::uint64_t size(const T (&)[N]) noexcept
+{
+    return N;
+}
+
+template <class C>
+constexpr auto data(C& c) -> decltype(c.data())
+{
+    return c.data();
+}
+
+template <class C>
+constexpr auto data(const C& c) -> decltype(c.data())
+{
+    return c.data();
+}
+
+template <class T, std::uint64_t N>
+constexpr T* data(T (&array)[N]) noexcept
+{
+    return array;
+}
+
+template <class E>
+constexpr const E* data(std::initializer_list<E> il) noexcept
+{
+    return il.begin();
+}
+
+template <typename T, uint64_t Extent = dynamic_extent>
```

Comment:
What is Extend used for? (I will probably find out later in the review).

Reply:
It is a requirement for `std::span`:

> A span can either have a static extent, in which case the number of elements in the sequence is known at compile-time and encoded in the type, or a dynamic extent

see [en.cppreference.com](https://en.cppreference.com/w/cpp/container/span)